### PR TITLE
Improve match like matches lint

### LINKED ISSUE #6186: match_like_matches_macro lint could be improved a bit more
For example, this case doesn't warn:

```rust
match self {
    WorkMode::Normal => true,
    WorkMode::Sys => true,
    _ => false,
}
```

But it could be improved using the `matches!` macro:

```rust
matches!(self, WorkMode::Normal | WorkMode::Sys)
```

### PR DESCRIPTION
**Pull request by alex-700**
fixes #6186 

changelog: improve MATCH_LIKE_MATCHES_MACRO lint

### CODE REVIEW

**Comment by alex-700** [comment]
@ebroto, I remember about [this](https://github.com/rust-lang/rust-clippy/issues/6186#issuecomment-711440305), but don't know how to do it without global refactoring near these two lints. Do you have some other ideas? Is there a big issue for clippy: don't lint two different lints for the same code?

**Comment by ebroto** [comment]
>but don't know how to do it without global refactoring near these two lints.

One approach is to move `match_same_arms` into the `matches` module where `match_like_matches_macro` lives, and not trigger the former if the latter lints. I just did something similar in [this commit](https://github.com/rust-lang/rust-clippy/pull/5911/commits/dceb43b78ae76886625e364c10da8917dce16251) for another lint.

In this case we would have to move some functions from the `copies` module to `utils`. Let me know if I explained my idea correctly :)

**Comment by ebroto** [comment]
>Is there a big issue for clippy: don't lint two different lints for the same code?

Giving two different suggestions for the same code can be misleading if they suggest different things, so we try to avoid it. Moreover, if both lints have auto-applicable suggestions, trying to apply the second one would fail.

**Comment by alex-700** [comment]
> Let me know if I explained my idea correctly

Yeap, I think that your idea is perfectly clear. It's the one I called "global refactoring near these two lints" :) The main problem that all examples in clippy code, I see, were straightforward methods, which call `cx.span_*` after "batch of ifs". For reusing the code for non-linting something I need some another approach. I'm choosing between the following:

- introduce method which will return `Option<SpanContext>`, instead of calling `cx.span_*`, where `SpanContext` is some information that this lint gathered during "batch of ifs". But what should be stored in this `SpanContext`?..
- like the first one, but returns `Option<impl Fn()>` or mb `Option<dyn Fn()>`, where call of this function is exactly "spanning the lint"
- have some set of spans of other lints, which can conflict with my one (like in your example). In this case I need to track the order of checks (introduce common lint-group for this one), also I need to share between several `fns` information about which of these spans I saved into the set. IMHO, this option isn't good for following code support.

@ebroto did you get, what I'm trying to say?  Is the third idea common for clippy codebase? If it is, can you point several others examples? Personally I like the second idea in cases where there is no some cpu-consuming calculations in this `Option<...>` evaluation.

File: tests/ui/match_expr_like_matches_macro.rs [hunk 1]
```diff
@@ -45,4 +45,54 @@ fn main() {
...
+        let _ans = match x {
+            E::A(_) => true,
+            E::B(_) => false,
+            E::C => false,
+            _ => true,
+        };
+    }
+    {
+        // no lint
+        let _ans = match x {
```
Comment (ebroto) [inline tests/ui/match_expr_like_matches_macro.rs:92]: Let's also add a test similar to this:
```rust
match x {
  E::A(_) => false,
  E::B(_) => true,
  _ => false,
}
```

**Review by ebroto — CHANGES_REQUESTED** [review]
The implementation LGTM, I've just suggested to add some tests.

About the interaction with `match_same_arms`, I apologize for not being more precise in my first comment. I pointed you to that commit to show that we can move a lint to another module, but the solution of storing the spans is very specific to that case because the conflicting lints start a different levels, one in `check_block` and the other one in `check_expr`.

In this case, I think it's easier because both lints work at the expression level, and if the expression is a match. So the solution should be applied in [this section of the `matches` module](https://github.com/rust-lang/rust-clippy/blob/afbac8906e614a63ff5825710c3ebe45a3b5e01a/clippy_lints/src/matches.rs#L508-L527) and doing something like
```rust
let matches_applied = check_match_like_matches(cx, expr);
// ...
if !matches_applied {
    check_match_same_arms(cx, expr);
}
```
I want to say that every time we try to avoid conflicting lints, it's a hack. We need a more general solution, probably implemented in rustc itself.

**Pull request merged**